SharpCaster/ChromeCastClient.cs

```diff
@@ -1,222 +1,272 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Diagnostics;
+﻿using Sharpcaster.Core.Channels;
+using Sharpcaster.Core.Interfaces;
+using Sharpcaster.Core.Models;
+using System;
 using System.IO;
-using System.Linq;
+using System.Net.Security;
+using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
 using Extensions.Api.CastChannel;
-using SharpCaster.Channels;
-using SharpCaster.Controllers;
-using SharpCaster.Extensions;
-using SharpCaster.Models;
-using SharpCaster.Models.ChromecastStatus;
-using SharpCaster.Models.MediaStatus;
-using SharpCaster.Interfaces;
-using SharpCaster.Services;
-
-namespace SharpCaster
+using static Extensions.Api.CastChannel.CastMessage.Types;
+using System.Text;
+using Newtonsoft.Json;
+using Sharpcaster.Core.Messages;
+using System.Collections.Generic;
+using System.Linq;
+using Sharpcaster.Extensions;
+using System.Collections.Concurrent;
+using System.Reflection;
+using Sharpcaster.Core.Models.ChromecastStatus;
+using System.Diagnostics;
+using Microsoft.Extensions.DependencyInjection;
+using Sharpcaster.Logging;
+
+namespace Sharpcaster
 {
-    public class ChromeCastClient
+    public class ChromecastClient : IChromecastClient
     {
-        public Volume Volume
+        private const int RECEIVE_TIMEOUT = 30000;
+        private static readonly object LockObject = new object();
+        private TcpClient _client;
+        private Stream _stream;
+        private ChromecastReceiver _receiver;
+        private TaskCompletionSource<bool> ReceiveTcs { get; set; }
+        public static readonly ILog Logger = LogProvider.For<ChromecastClient>();
+        private SemaphoreSlim SendSemaphoreSlim { get; } = new SemaphoreSlim(1, 1);
+
+        private IDictionary<string, Type> MessageTypes { get; set; }
+        private IEnumerable<IChromecastChannel> Channels { get; set; }
+        private ConcurrentDictionary<int, object> WaitingTasks { get; } = new ConcurrentDictionary<int, object>();
+
+        public ChromecastClient()
```

Comment:
Looks really interesting and I think that totally makes sense. I'm currently not happy with the constructor or the way it does DI. I haven't really looked into it that much yet. Once I get into the state where I create the custom channel for Web page casting I will look into changing it.

My current priority list:

1. Media Channel
2. More test cases (Disconnection, Error, Retry and etc)
3. Test multiple chromecasts at once
4. Building new channels as extensions